ONSdigital/dis-wagtail
# Introduce Functional Tests using Playwright and Behave

### What is the context of this PR?

This PR adds functional tests for the app, written in the Behave BDD framework and using Playwright to test browser interactions.

### How to review

- Try running the functional tests in your development environment, following the new instructions in the README

- Review the new documentation. Does it provide enough guidance for you to be happy adding new test features/scenarios as part of future work? Do the suggested code standards and patterns make sense? These are intended to be a starting point, based on my experience with large test suites in Behave, but they should be reviewed agreed as a team, not just taken as written. 

- Try running the tests in your IDE or editor of choice. Is it possible to easily run individual scenarios or debug them? Certain functionality might require installing extensions.

- Check the custom fixture code for test setup and teardown, does it all make sense and is it sufficiently documented and understandable?

- Check the features and scenarios. These are my initial guesses at some suitable scenarios, they are not driven by any explicit requirements from those features. I am perfectly happy to modify them, or even tear them up and do it differently if people take issue, I just needed some example features in place to demonstrate the code structure.

### Follow-up Actions

We still need to agree how we will utilise this method on testing in further development. I would suggest the test scenarios need to be agreed at refinement time for future incoming tickets, but this still needs wider discussion and agreement around how these new tests fit into our definitions or ready and done.

We will also need to dicuss and add tasks for backfilling tests on existing features.

Topics to Revisit:
- Potential to use `serialized_rollback` instead of our hand-rolled DSLR based database clean/restore between tests. Can we get it working and to a satisfactory speed?
- Building reusable page classes for consistent navigation in playwright
- Consider setting `IS_EXTERNAL_ENV` to `true` when testing the external site, to better replicate a production setting
- Pursuing multi-browser testing

### Known Issues

- The initial scenarios currently only reliably pass in the `chromium` browser. We have agreed to address this in future work.
